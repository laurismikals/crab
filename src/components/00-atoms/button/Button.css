.btn {
  display: inline-block;
  border-style: solid;
  border-width: 0;
  border-color: #000000;
  border-radius: 0;
  cursor: pointer;
  position: relative;
  overflow: hidden;
  user-select: none;
  text-align: center;
  transition:
      color 0.3s,
      background-color 0.3s,
      border-color 0.3s,
      opacity 0.3s;
  background-clip: border-box;
  background-color: transparent;
  text-decoration: none;
  color:inherit;
  height: 2.5em;
  min-width: 2.5em;
  padding: 10px 40px;
}
@media print {
  .btn {
    display: none;
  }
}
.btn__flex-fix {
  /* In Firefox button doesn't respect flexbox declaration. Inserting child element and setting display to flex fixes the issue. */
  width: 100%;
  height: 100%;
  position: relative;
  flex: 1 1 auto;
  display: flex;
  flex-wrap: nowrap;
  flex-direction: row;
  justify-content: center;
  align-items: center;
}

.btn--large {
  height: 4.000em;
  min-width: 4.000em;
}

.btn--primary {
  color: #ffffff;
  background-color: rgba(58, 199, 58, 1);
  border-color: rgba(58, 199, 58, 1);
  outline: none;
}
.btn--primary:hover,
.btn--primary:focus{
  color: #ffffff;
  background-color: rgba(58, 199, 58, 0.7);
  border-color: rgba(58, 199, 58, 0.7);
}
.btn--primary:active,
.btn--primary.active,
.btn--primary[aria-selected=true] {
  color: #ffffff;
  background-color: rgba(58, 199, 58, 0.5);
  border-color: rgba(58, 199, 58, 0.5);
}

.btn--disabled,
.btn[disabled] {
  color: #cccccc;
}

.btn__text {
  width: 100%;
  position: relative;
  text-align: center;
  font-weight: 700;
  line-height: 1.3;
  white-space: nowrap;
  overflow: hidden;
  flex: 1 1 auto;
}

.btn--loading{
  padding-right: 40px;
}
.btn--loading:after{
  content: '';
  position: absolute;
  top: 50%;
  left: 100%;
  width: 18px;
  height: 18px;
  margin: -9px 0 0 -30px;
  border-style: solid;
  border-bottom-color: transparent;
  border-width: 2px;
  border-radius: 50%;
  animation: loadingSpin 1s infinite linear;
}
@keyframes loadingSpin{
  from{
    transform: rotate(0deg);
  }
  to{
    transform: rotate(360deg);
  }
}
