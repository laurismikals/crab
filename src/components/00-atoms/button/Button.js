import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import includes from 'lodash/fp/includes';

import './Button.css';

export class Button extends PureComponent{
  static propTypes = {
    type: (props, propName, componentName) => {
      if (!includes(props[propName])(['button', 'submit', 'reset'])) {
        return new Error(
          'Invalid prop `' + propName + '` supplied to' +
          ' `' + componentName + '`. Validation failed.' +
          'It should be one of the following: button, submit, reset.'
        );
      }
    },
    size: (props, propName, componentName) => {
      if (!includes(props[propName])(['small', 'large'])) {
        return new Error(
          'Invalid prop `' + propName + '` supplied to' +
          ' `' + componentName + '`. Validation failed.' +
          'It should be one of the following: small, large.'
        );
      }
    },
    appearance: (props, propName, componentName) => {
      if (!includes(props[propName])(['default', 'primary'])) {
        return new Error(
          'Invalid prop `' + propName + '` supplied to' +
          ' `' + componentName + '`. Validation failed.' +
          'It should be one of the following: default, primary.'
        );
      }
    },
    text: PropTypes.string.isRequired,
    className: PropTypes.string,
  };
  static defaultProps = {
    type: 'button',
    className: '',
  };
  render() {
    const {
      type,
      size,
      appearance,
      className,
      text,
      ...others
    } = this.props;

    return (
      <button
        type={type}
        className={`
          btn
          ${className}
          btn--${size}
          btn--${appearance}
        `}
        {...others}
      >
        <span className="btn__flex-fix">
          <span className="btn__text">{text}</span>
        </span>
      </button>
    )
  }
}

export default Button;
