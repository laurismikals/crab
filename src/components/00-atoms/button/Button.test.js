import React from 'react';
import Enzyme, { shallow } from 'enzyme';

import { Button } from './Button';

describe('Atom: Button', () => {
  const minProps = {
    size: 'large',
    appearance: 'default',
    text: 'Button text',
  };

  it('renders correctly', () => {
    const tree = shallow(
      <Button {...minProps} />
    );
    expect(tree).toMatchSnapshot();
  });
});
