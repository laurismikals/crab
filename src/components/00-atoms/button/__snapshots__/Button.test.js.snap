// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Atom: Button renders correctly 1`] = `
<button
  className="
          btn
          
          btn--large
          btn--default
        "
  type="button"
>
  <span
    className="btn__flex-fix"
  >
    <span
      className="btn__text"
    >
      Button text
    </span>
  </span>
</button>
`;
